Add support for Cisco-8000 400G

rbpittman opened the pull request:
<!--
Please make sure you've read and understood our contributing guidelines;
https://github.com/sonic-net/SONiC/blob/gh-pages/CONTRIBUTING.md

Please provide following information to help code review process a bit easier:
-->
### Description of PR
<!--
- Please include a summary of the change and which issue is fixed.
- Please also include relevant motivation and context. Where should reviewer start? background context?
- List any dependencies that are required for this change.
-->

Add support for Cisco-8000 400G platform.

Summary:
Fixes # (issue)

### Type of change

<!--
- Fill x for your type of change.
- e.g.
- [x] Bug fix
-->

- [ ] Bug fix
- [x] Testbed and Framework(new/improvement)
- [x] Test case(new/improvement)


### Back port request
- [ ] 201911
- [ ] 202012
- [x] 202205

### Approach
#### What is the motivation for this PR?

Support Cisco-8000 QOS SAI test cases for 400G.

#### How did you do it?
Additional qos.yml entries tuned for 400G.

Revised shared reservation size test case to use port indices instead of IDs, allows for more generalized and reliable testing across platforms.

Shared res size now accounts for destinations that may be in a port-channel.

Py3 adaptations for some tests.

Some formatting changes during execution of `pre-commit`.

#### How did you verify/test it?
Verified test cases internally, all pass except:

Flaky:
testQosSaiPfcXoffLimit[dut-400g-None-xoff_2]

Fails, not cleaned up yet for 400G platform:
testQosSaiLosslessVoq
testQosSaiLossyQueueVoq

#### Any platform specific information?

#### Supported testbed topology if it's a new test case?

### Documentation
<!--
(If it's a new feature, new test case)
Did you update documentation/Wiki relevant to your implementation?
Link to the wiki page?
-->

kevinskwang commented:
Will this impact the 8102?

neethajohn commented on tests/qos/files/qos.yml:
```diff
@@ -2987,8 +2987,8 @@ qos_params:
                     queues: [0, 0, 0, 0, 0, 3, 4, 3, 4, 3, 4, 3, 4, 3, 4, 3, 4]
-                    src_port_ids: [0, 4, 16, 20, 34, 0, 0, 4, 4, 16, 16, 20, 20, 34, 34, 36, 36]
-                    dst_port_ids: [37, 38, 39, 42, 44, 37, 37, 38, 38, 39, 39, 42, 42, 44, 44, 45, 45]
+                    src_port_ids: [0, 1, 2, 3, 4, 0, 0, 1, 1, 2, 2, 3, 3, 4, 4, 5, 5]
```
Why are these changed? These were the configs for 100G

rbpittman replied:
The test has been uplifted as part of the new platform support for 400G. It is now agnostic across port layouts. Instead of these IDs being unique port IDs, they're now indices that the test parses into whatever valid ports the testbed has. Thus the 100G test needed to receive this config change as well.

rbpittman replied:
I went ahead and made an internal preview of the 202012 branch double commit for these tests, and have confirmed that these changes do not affect the tests.

rbpittman commented:
> Will this impact the 8102?

Yes it will be running different code, so verification is helpful here. 
Since the 8102 is running 202012, I've internally previewed the future double commit of this to 202012 and confirmed the test behavior remains the same.

kevinskwang commented:
> > Will this impact the 8102?
> 
> Yes it will be running different code, so verification is helpful here. Since the 8102 is running 202012, I've internally previewed the future double commit of this to 202012 and confirmed the test behavior remains the same.

There is a plan to shift 8102 to 202205. So we need to consider that as well. 
Please also fix the conflicts.

rbpittman commented:
Looking forward to using 202205 instead of 202012.
The new conflict that showed up appears to change some important port ID infra. I'll need to do some retesting with the resolution.

wangxin commented:
The [pre-commit](http://pre-commit.com/) check detected issues in the files touched by this pull request.
The detected issues may be old or new. For new issues, please try to fix them.

For old issues, it is not mandatory to fix them because they were not caused by this change. It is unfair to blame
author of this pull request. But if you can take extra effort to fix the old issues as well, that would be great!

Detailed pre-commit check results:
<samp>trim trailing whitespace.................................................Failed<br/>- hook id: trailing-whitespace<br/>- exit code: 1<br/>- files were modified by this hook<br/><br/>Fixing tests/qos/files/qos.yml<br/>Fixing tests/qos/test_qos_sai.py<br/><br/>fix end of files.........................................................Failed<br/>- hook id: end-of-file-fixer<br/>- exit code: 1<br/>- files were modified by this hook<br/><br/>Fixing tests/saitests/py3/sai_qos_tests.py<br/><br/>check yaml...............................................................Passed<br/>check for added large files..............................................Passed<br/>check python ast.........................................................Passed<br/>flake8...................................................................Failed<br/>- hook id: flake8<br/>...<br/>[truncated extra lines, please run pre-commit locally to view full check results]<br/></samp>

To run the pre-commit checks locally, you can follow below steps:
1. Ensure that default python is python3. In sonic-mgmt docker container, default python is python2. You can run
   the check by activating the python3 virtual environment in sonic-mgmt docker container or outside of sonic-mgmt
   docker container.
2. Ensure that the `pre-commit` package is installed:
```
sudo pip install pre-commit
```
3. Go to repository root folder
4. Install the pre-commit hooks:
```
pre-commit install
```
5. Use pre-commit to check staged file:
```
pre-commit
```
6. Alternatively, you can check committed files using:
```
pre-commit run --from-ref <commit_id> --to-ref <commit_id>
```

rbpittman commented:
The new port ID checker infra is now interfering with the test. Working on a fix.

wangxin commented:
The [pre-commit](http://pre-commit.com/) check detected issues in the files touched by this pull request.
The detected issues may be old or new. For new issues, please try to fix them.

For old issues, it is not mandatory to fix them because they were not caused by this change. It is unfair to blame
author of this pull request. But if you can take extra effort to fix the old issues as well, that would be great!

Detailed pre-commit check results:
<samp>trim trailing whitespace.................................................Failed<br/>- hook id: trailing-whitespace<br/>- exit code: 1<br/>- files were modified by this hook<br/><br/>Fixing tests/qos/files/qos.yml<br/>Fixing tests/qos/test_qos_sai.py<br/>Fixing tests/saitests/py3/sai_qos_tests.py<br/><br/>fix end of files.........................................................Failed<br/>- hook id: end-of-file-fixer<br/>- exit code: 1<br/>- files were modified by this hook<br/><br/>Fixing tests/saitests/py3/sai_qos_tests.py<br/><br/>check yaml...............................................................Passed<br/>check for added large files..............................................Passed<br/>check python ast.........................................................Passed<br/>flake8...................................................................Failed<br/>...<br/>[truncated extra lines, please run pre-commit locally to view full check results]<br/></samp>

To run the pre-commit checks locally, you can follow below steps:
1. Ensure that default python is python3. In sonic-mgmt docker container, default python is python2. You can run
   the check by activating the python3 virtual environment in sonic-mgmt docker container or outside of sonic-mgmt
   docker container.
2. Ensure that the `pre-commit` package is installed:
```
sudo pip install pre-commit
```
3. Go to repository root folder
4. Install the pre-commit hooks:
```
pre-commit install
```
5. Use pre-commit to check staged file:
```
pre-commit
```
6. Alternatively, you can check committed files using:
```
pre-commit run --from-ref <commit_id> --to-ref <commit_id>
```

rbpittman commented:
@kevinskwang @neethajohn 
I've finished polishing up the code after that recent set of conflicts. The new port ID checker is interesting, but does not apply to this test since I've adapted it to use "indices", such that the IDs can be safely and automatically deduced across platforms. I renamed the qos.yml to avoid the new infra's regex from picking this up and thinking it's an ID it needs to correct.

I also removed the excess diff in other tests to keep it short. Those diffs are for the multi-flow test cases which are in the second upcoming PR.

wangxin commented:
@rbpittman There is conflict while cherry-picking to 202205 branch. Can you help create separate PR to the 202205 branch to include this change?

rbpittman commented:
Sure thing @wangxin.

I've filed #7653. 

The resolution looked good as a "theirs" selection, if it needs any fixes it'd probably not be appropriate to include here, as this should be a completely clean "double-commit".
